myia/prim/type_inferrers.py

```diff
@@ -199,6 +199,75 @@ def default(self):
     return t
 
 
+@type_inferrer(P.shape, nargs=1)
+async def infer_type_shape(engine, ary):
+    """Infer the return type of shape."""
+    shp = await engine.get('shape', ary)
+    return Tuple([Int(64)]*len(shp))
+
+
+@type_inferrer(P.map_array, nargs=2)
+async def infer_type_map_array(engine, fn, ary):
+    """Infer the return type of map_array."""
+    fn_t = await engine.get('type', fn)
+    ary_t = await engine.get('type', ary)
+    if not isinstance(ary_t, Array):
+        raise MyiaTypeError('Expected array')
+    xref = engine.vref({'type': ary_t.elements})
+    return Array(await fn_t(xref))
+
+
+@type_inferrer(P.scan_array, P.reduce_array, nargs=4)
+async def infer_type_across_array(engine, fn, init, ary, ax):
+    """Infer the return type of map_array."""
+    fn_t = await engine.get('type', fn)
+    ary_t = await engine.get('type', ary)
+    init_t = await engine.get('type', init)
+    ax_t = await engine.get('type', ax)
+    if not isinstance(ary_t, Array):
+        raise MyiaTypeError('Expected array')
+    if not ary_t.elements == init_t:
+        raise MyiaTypeError("Initial value must have the same type "
+                            "as array elements")
+    if not ax_t == UInt(64):
+        raise MyiaTypeError("Axis must be u64")
```

Comment:
Do we even have any way to create a `UInt(64)` at the moment?

Reply:
Not with the parser.